src/compiler/checker.ts

```diff
@@ -1923,32 +1923,42 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function runWithoutResolvedSignatureCaching<T>(node: Node | undefined, fn: () => T): T {
+        const cachedResolvedSignatures = [];
+        const cachedTypes: (readonly [SymbolLinks, Type | undefined])[] = [];
         node = findAncestor(node, isCallLikeOrFunctionLikeExpression);
-        if (node) {
-            const cachedResolvedSignatures = [];
-            const cachedTypes = [];
-            while (node) {
-                const nodeLinks = getNodeLinks(node);
-                cachedResolvedSignatures.push([nodeLinks, nodeLinks.resolvedSignature] as const);
-                nodeLinks.resolvedSignature = undefined;
-                if (isFunctionExpressionOrArrowFunction(node)) {
-                    const symbolLinks = getSymbolLinks(getSymbolOfDeclaration(node));
-                    const type = symbolLinks.type;
-                    cachedTypes.push([symbolLinks, type] as const);
-                    symbolLinks.type = undefined;
-                }
-                node = findAncestor(node.parent, isCallLikeOrFunctionLikeExpression);
-            }
-            const result = fn();
-            for (const [nodeLinks, resolvedSignature] of cachedResolvedSignatures) {
-                nodeLinks.resolvedSignature = resolvedSignature;
-            }
-            for (const [symbolLinks, type] of cachedTypes) {
-                symbolLinks.type = type;
+        if (!node) {
+            return fn();
+        }
+        while (node) {
+            const nodeLinks = getNodeLinks(node);
+            cachedResolvedSignatures.push([nodeLinks, nodeLinks.resolvedSignature, nodeLinks.flags] as const);
+            nodeLinks.resolvedSignature = undefined;
+            nodeLinks.flags = NodeCheckFlags.None;

```

Comment:
The key here is to reset `NodeCheckFlags.ContextChecked` so `contextuallyCheckFunctionExpressionOrObjectLiteralMethod` can correctly resolve (and cache) `signature.resolvedReturnType` multiple times (for requests with inference blocked and for requests without inference blocked). Otherwise, it has to be resolved later on... but without proper contextual information.

So I could just remove that bit and re-add it later if needed. I looked through all `NodeCheckFlags` and it might also be needed to do the same for `NodeCheckFlags.TypeChecked`. Although, I'm not 100% sure if that's really the case.

I just tried with a complete flags reset here and it does seem to work... but 🤷‍♂️ I'd probably be in favor of resetting solely what's really-really needed.